pkg/rancher/migrations.go

```diff
@@ -624,6 +634,68 @@ func migrateSystemAgentDataDirectory(w *wrangler.Context) error {
 	return createOrUpdateConfigMap(w.Core.ConfigMap(), cm)
 }
 
+func migrateHarvesterCloudCredentialExpiration(w *wrangler.Context) error {
+	cm, err := getConfigMap(w.Core.ConfigMap(), migrateHarvesterCloudCredentialExpirationConfig)
+	if err != nil || cm == nil {
+		return err
+	}
+
+	if cm.Data[harvesterCloudCredentialExpirationMigratedKey] == "true" {
+		return nil
+	}
+
+	secrets, err := w.Core.Secret().List("cattle-global-data", metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	for _, secret := range secrets.Items {
+		if content, ok := secret.Data["harvestercredentialConfig-kubeconfigContent"]; ok && content != nil {
+			kubeConfig := map[string]any{}
+			if err = yaml.Unmarshal(content, &kubeConfig); err != nil {
+				return err
+			}
+
+			tokenName := ""
+			if userList, ok := kubeConfig["users"].([]any); ok && userList != nil && len(userList) > 0 {
+				for _, u := range userList {
+					if entry, ok := u.(map[string]any); ok && entry != nil {
+						if user, ok := entry["user"].(map[string]any); ok && user != nil {
+							if token, ok := user["token"].(string); ok && token != "" {
+								if strings.HasPrefix(tokenName, "kubeconfig-user-") {
+									tokenName, _, _ = strings.Cut(tokenName, ":")
+									break
+								}
+							}
+						}
+					}
+				}
+			}
```

Comment:
imo we should have this common logic in a func, we're using the same thing 3 times

Reply:
I've refactored the logic into 4 separate functions, one public which is also used in migrations.go which has an anonymous function parameter for actually getting the token from the API. I've yet to add comments or tests, but if you are okay with the general approach let me know and I will do so.

Reply:
yep, I like this :)